File: contracts/atomic-nft-typescript/lib/balance.ts
import { AtomicNFTState } from "./faces"
import { isAddress, Result } from "./utils"

export type BalanceOfResult = {
    result: {
        balance: string,
        ticker: string,
        target: string
    }
}

/**
 * Returns the amount of tokens owned by `target`.
 * @param state this contract mutable state
 * @param target
 * @returns balance of `target`
 */
export function balanceOf(state: AtomicNFTState, target: string) {
    isAddress(target, "target");

    return Result({
        balance: state.balances[target] ?? 0,
        ticker: state.symbol,
        target
    })
}

export type TotalSupplyResult = {
    result: {
        value: number
    }
}

/**
 * Returns the amount of tokens in existence.
 * @param state this contract mutable state
 * @returns the amount of tokens in existence.
 */
export function totalSupply(state: AtomicNFTState): TotalSupplyResult {
    return Result({
        value: state.totalSupply
    })
}

File: contracts/erc20/src/actions/balance.rs
use crate::{
    action::{
        QueryResponseMsg::{Balance, TotalSupply},
        ViewResponse,
    },
    state::State,
};

pub fn balance_of(state: &State, target: String) -> ViewResponse {
    ViewResponse::Success(Balance {
        balance: *state.balances.get(&target).unwrap_or(&0),
        ticker: state.symbol.clone(),
        target,
    })
}

pub fn total_supply(state: &State) -> ViewResponse {
    ViewResponse::Success(TotalSupply {
        value: state.total_supply,
    })
}
